Improve helm documentation using helm-docs and markdown-toc (#59)
voxel51/fiftyone-teams-app-deploy: helm/fiftyone-teams-app/README.md

    - [Create a ClusterIssuer](#create-a-clusterissuer)
    - [Install and Configure MongoDB](#install-and-configure-mongodb)
    - [Obtain a Global Static IP Address and Configure a DNS Entry](#obtain-a-global-static-ip-address-and-configure-a-dns-entry)
    - [Set up http to https Forwarding](#set-up-http-to-https-forwarding)
    - [Install FiftyOne Teams App](#install-fiftyone-teams-app)
    - [Installation Complete](#installation-complete)

<!-- tocstop -->

We publish container images to these Docker Hub repositories

- `voxel51/fiftyone-app`
- `voxel51/fiftyone-app-gpt`
- `voxel51/fiftyone-app-torch`
- `voxel51/fiftyone-teams-api`
- `voxel51/fiftyone-teams-app`

For Docker Hub credentials, please contact your Voxel51 support team.

## Initial Installation vs. Upgrades

Upgrades are more frequent than new installations.
Thus, the chart's default behavior supports
upgrades and the `values.yaml` contains

```yaml
appSettings:
  env:
    FIFTYONE_DATABASE_ADMIN: false
```

When performing an initial installation,
in your `values.yaml`, set

```yaml
appSettings:
  env:
    FIFTYONE_DATABASE_ADMIN: true
```

> After the initial installation, either comment this
> environment variable or change the value to false.

Reviewer: ```suggestion
After the initial installation, we recommend either
commenting this environment variable or changing
the value to false.
```

> When performing an upgrade, please review our

Reviewer: ```suggestion
When performing an upgrade, please review
```

[Upgrading From Previous Versions](#upgrading-from-previous-versions)

## FiftyOne Features

Consider if you will require these settings for your deployment.

### FiftyOne Teams Authenticated API

FiftyOne Teams v1.3 introduced the capability to connect FiftyOne Teams SDKs
through the FiftyOne Teams API (instead of direct MongoDB connection).

To enable the FiftyOne Teams Authenticated API,

> [expose the FiftyOne Teams API endpoint](docs/expose-teams-api.md)

Reviewer: ```suggestion
[expose the FiftyOne Teams API endpoint](../docs/expose-teams-api.md)
```

Reviewer: Moving the README, but not the docs directory, means only part of the docs will ship with the helm chart.  Do you need to move the `docs` directory here too?

Author: If you think we should include it, we can. Let me know what you'd like.

Reviewer: well, we probably need to either update the links to be fully-qualified pointing at helm.fiftyone.ai or bundle them - otherwise the links won't work...

Either way, doesn't really matter to me

Author: I'll update the links to be :

* https://github.com/voxel51/fiftyone-teams-app-deploy/blob/main/helm/docs/expose-teams-api.md
* https://github.com/voxel51/fiftyone-teams-app-deploy/blob/main/helm/docs/plugins-storage.md

That way, whether viewed in GH or in the chart's README.md, they always work.

and
[configure your SDK](https://docs.voxel51.com/teams/api_connection.html).

### FiftyOne Teams Plugins

FiftyOne Teams v1.3 introduced significant enhancements for
[Plugins](https://docs.voxel51.com/plugins/index.html)
to customize and enhance functionality.

There are three modes for plugins

1. Builtin Plugins Only
    - No changes are required for this mode
1. Plugins run in the `fiftyone-app` deployment
    - To enable this mode
        - In `values.yaml`, set the path for a Persistent Volume Claim
          mounted to the `teams-api` and `fiftyone-app` deployments in both
            - `appSettings.env.FIFTYONE_PLUGINS_DIR`
            - `apiSettings.env.FIFTYONE_PLUGINS_DIR`

>         - Mount a [Persistent Volume Claim](docs/plugins-storage.md) that provides

Reviewer: ```suggestion
        - Mount a [Persistent Volume Claim](../docs/plugins-storage.md) that provides
```

            - `ReadWrite` permissions to the `teams-api` deployment
              at the `FIFTYONE_PLUGINS_DIR` path
            - `ReadOnly` permission to the `fiftyone-app` deployment
              at the `FIFTYONE_PLUGINS_DIR` path
1. Plugins run in a dedicated `teams-plugins` deployment
    - To enable this mode
        - In `values.yaml`, set
            - `pluginsSettings.enabled: true`
            - The path for a Persistent Volume Claim mounted to the
              `teams-api` and `teams-plugins` deployments in both
                - `pluginsSettings.env.FIFTYONE_PLUGINS_DIR`
                - `apiSettings.env.FIFTYONE_PLUGINS_DIR`

>         - Mount a [Persistent Volume Claim](docs/plugins-storage.md) that provides

Reviewer: ```suggestion
        - Mount a [Persistent Volume Claim](../docs/plugins-storage.md) that provides
```

            - `ReadWrite` permissions to the `teams-api` deployment
              at the `FIFTYONE_PLUGINS_DIR` path
            - `ReadOnly` permission to the `teams-plugins` deployment
              at the `FIFTYONE_PLUGINS_DIR` path

Use the FiftyOne Teams UI to deploy plugins by navigating to `https://<DEPOY_URL>/settings/plugins`.
Early-adopter plugins installed manually must be
redeployed using the FiftyOne Teams UI.

### Storage Credentials and `FIFTYONE_ENCRYPTION_KEY`

Pods based on the `fiftyone-teams-api` and `fiftyone-app`
images must include the `FIFTYONE_ENCRYPTION_KEY` variable.
This key is used to encrypt storage credentials in the MongoDB database.

The generate an `encryptionKey`, run this Python code

```python
from cryptography.fernet import Fernet
print(Fernet.generate_key().decode())
```

Voxel51 does not have access to this encryption key and cannot reproduce it.
Please store this key in a safe place.
If the key is lost, you will need to

1. Schedule an outage window
    1. Drop the storage credentials collection
    1. Replace the encryption key
    1. Add the storage credentials via the UI again.

> Storage credentials no longer need to be mounted into
> pods with appropriate environment variables being set.

Reviewer: ```suggestion
```

Users with `Admin` permissions may use the FiftyOne Teams UI to manage storage
credentials by navigating to `https://<DEPOY_URL>/settings/cloud_storage_credentials`.

> 

Reviewer: ```suggestion
If added via the UI, storage credentials no
longer need to be mounted into pods or 
provided via environment variables.

```

    ```

1. The pod based on the `fiftyone-teams-app` image (`teamsAppSettings.env`)

    ```yaml
    GLOBAL_AGENT_HTTP_PROXY: http://proxy.yourcompany.tld:3128
    GLOBAL_AGENT_HTTPS_PROXY: https://proxy.yourconpay.tld:3128
    GLOBAL_AGENT_NO_PROXY: <apiSettings.service.name>, <appSettings.service.name>, <teamsAppSettings.service.name>
    ```

The `NO_PROXY` and `GLOBAL_AGENT_NO_PROXY` values must include the Kubernetes
service names that may communicate without going through a proxy server.
By default, these service names are

- `teams-api`
- `teams-app`
- `fiftyone-app`

If the service names were overridden in `*.service.name`, use these values instead.

By default, the Global Agent Proxy will log all outbound connections
and identify which connections are routed through the proxy.
To reduce the logging verbosity, add this environment variable to your `teamsAppSettings.env`

```ini
ROARR_LOG: false
```

#### Text Similarity

Since version v1.2, FiftyOne Teams supports using text similarity
searches for images that are indexed with a model that
[supports text queries](https://docs.voxel51.com/user_guide/brain.html#brain-similarity-text).

> To use this feature, use a container image containing
> `torch` (PyTorch), instead of the `fiftyone-app` image.

Reviewer: ```suggestion
```

Reviewer: This statement is confusing to me, because it fails to note that you need all the same packages as `fiftyone-app` _and_ PyTorch...  I say drop it and just use the sentence below.

> Use the Voxel51 provided image `fiftyone-app-torch` or
> build your own base image including `torch`.

Reviewer: ```suggestion
Use the Voxel51 provided image `fiftyone-app-torch` or
build your own base image including `torch` (PyTorch).
```

| teamsAppSettings.service.name | string | `"teams-app"` | Service name. |
| teamsAppSettings.service.nodePort | int | `nil` | Service nodePort set only when `teamsAppSettings.service.type: NodePort` for teams-app. |
| teamsAppSettings.service.port | int | `80` | Service port. |
| teamsAppSettings.service.readiness.initialDelaySeconds | int | `45` | Number of seconds to wait before performing the readiness probe for teams-app. [Reference][probes]. |
| teamsAppSettings.service.shortname | string | `"teams-app"` | Port name (maximum length is 15 characters) for teams-app. [Reference][ports]. |
| teamsAppSettings.service.type | string | `"ClusterIP"` | Service type for teams-app. [Reference][service-type]. |
| teamsAppSettings.tolerations | list | `[]` | Allow the k8s scheduler to schedule teams-app pods with matching taints. [Reference][taints-and-tolerations]. |
| teamsAppSettings.volumeMounts | list | `[]` | Volume mounts for teams-app pods. [Reference][volumes]. |
| teamsAppSettings.volumes | list | `[]` | Volumes for teams-app pods. [Reference][volumes]. |

## Upgrading From Previous Versions

### From Early Adopter Versions (Versions less than 1.0)

> You will need to either create a new Identity Provider (IdP) or modify your existing
> configuration to migrate to a new Auth0 Tenant.

Reviewer: ```suggestion
Please contact your Voxel51 Customer Success team
member to coordinate this upgrade.
You will need to either create a new Identity Provider (IdP) or modify your existing
configuration to migrate to a new Auth0 Tenant.
```

Reviewer: folks can't do this by themselves, they need us to make the Auth0 changes...
